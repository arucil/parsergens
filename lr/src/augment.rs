use grammar::NonterminalIdGen;
use grammar::{
  LoweredGrammar, Production, ProductionKind, Symbol,
  LoweredNonterminal, NonterminalKind,
};

/// Add `S' -> S $` to grammar, where `$` is a new token representing EOF.
pub fn augment(grammar: LoweredGrammar) -> LoweredGrammar {
  let max_nt_id = grammar.nts
    .keys()
    .map(|x| x.id())
    .max()
    .unwrap();
  let mut nt_id_gen = NonterminalIdGen::from(max_nt_id);

  let mut prods = grammar.prods;
  let mut nts = grammar.nts;

  let start_nts = grammar.start_nts.into_iter().map(|nt| {
    let new_start_nt = nt_id_gen.gen();
    let start = prods.len();

    prods.push(Production {
      nt: new_start_nt,
      kind: ProductionKind::Ordinary,
      symbols: vec![
        Symbol::Nonterminal(nt),
        //Symbol::Token(eof_token),
      ],
      prec: None,
      action: None,
    });

    let new_nt_name = format!("_{}", nts[&nt].name);

    nts.insert(new_start_nt, LoweredNonterminal {
      name: new_nt_name,
      range: start..prods.len(),
      ty: None,
      kind: NonterminalKind::User,
    });

    new_start_nt
  }).collect();

  let grammar = LoweredGrammar {
    prods,
    start_nts,
    nts,
    ..grammar
  };

  grammar
}
