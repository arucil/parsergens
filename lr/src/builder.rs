use std::collections::VecDeque;
use grammar::{LoweredGrammar, Symbol, Map, NonterminalId, HashMap, Assoc};
use indexmap::IndexMap;
use fnv::FnvBuildHasher;
use std::fmt::{self, Write};
use std::hash::Hash;
use crate::first::NonterminalFirst;
use crate::{
  EntryPoint, Error, ShiftReduceConflictError, ReduceReduceConflictError
};
use crate::token_set::TokenSet;

pub struct Builder<'a, T: LrComputation> {
  pub grammar: &'a LoweredGrammar,
  pub states: StateStore<T::StateKey>,
  /// eof is the token with the greatest id
  pub eof: usize,
  /// max number of RHS symbols in productions, plus one.
  pub max_nsym_p1: usize,
}

pub type StateStore<Key> = IndexMap<Key, State, FnvBuildHasher>;

/// sorted.
pub type KernelItemSet = Vec<Item>;

pub type LookaheadSet = TokenSet;

#[derive(Debug)]
pub struct State {
  /// Starts with sorted `kernel_len` kernel items, followed by other non-kernel items.
  pub items: Vec<Item>,
  /// size of kernel item set
  pub kernel_len: usize,
  /// symbol -> index of destination state
  pub transitions: Map<Symbol, u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Item {
  /// production and dot
  pub key: u32,
  pub lookaheads: LookaheadSet,
}

pub trait LrComputation {
  type StateKey: Eq + Hash;

  /// Returns the index of the state and whether the state has changed.
  fn store_state(
    states: &mut StateStore<Self::StateKey>,
    kernel_item_set: KernelItemSet,
  ) -> (u32, bool);
}

pub fn gen_states<T: LrComputation>(
  builder: &mut Builder<T>,
) -> HashMap<String, EntryPoint> {
  let nt_firsts = crate::first::compute(builder.grammar);
  let mut entry_points = HashMap::default();

  for &start_nt in &builder.grammar.start_nts {
    let start_state = start(builder, &nt_firsts, start_nt);
    let start_prod_ix = builder.grammar.nts[&start_nt].range.start;
    let real_start_nt =
      match builder.grammar.prods[start_prod_ix].symbols[0] {
        Symbol::Nonterminal(nt) => nt,
        _ => unreachable!(),
      };
    let nt_name = builder.grammar.nts[&real_start_nt].name.clone();
    entry_points.insert(nt_name, EntryPoint {
      real_start_nt: real_start_nt.id(),
      start_state,
      accept_prod: start_prod_ix,
    });
  }

  entry_points
}

fn start<T: LrComputation>(
  builder: &mut Builder<T>,
  nt_firsts: &[NonterminalFirst],
  start_nt: NonterminalId,
) -> u32 {
  let start_prod = builder.grammar.nts[&start_nt].range.start;
  let start_item_set = vec![
    Item {
      key: encode_item(builder.max_nsym_p1, start_prod, 0),
      lookaheads: TokenSet::from_token(builder.eof + 1, builder.eof as u32),
    }
  ];

  let (start_state, _) = T::store_state(&mut builder.states, start_item_set);

  let mut queue = VecDeque::new();
  queue.push_back(start_state);

  while let Some(state_ix) = queue.pop_front() {
    let state = &mut builder.states[state_ix as usize];
    compute_closure(builder.grammar, nt_firsts, builder.max_nsym_p1, state);

    let transitions = compute_transitions(builder.grammar, builder.max_nsym_p1, state);
    for (sym, mut kernel_item_set) in transitions {
      kernel_item_set.sort_by_key(|item| item.key);

      let (next_state, changed) = T::store_state(&mut builder.states, kernel_item_set);
      if changed {
        queue.push_back(next_state);
      }
      builder.states[state_ix as usize].transitions.insert(sym, next_state);
    }
  }

  start_state
}

fn compute_closure(
  grammar: &LoweredGrammar,
  nt_firsts: &[NonterminalFirst],
  max_nsym_p1: usize,
  state: &mut State,
) {
  let items = &mut state.items;
  // nt -> start index of items
  let mut nt_starts = HashMap::default();
  let mut first = TokenSet::new(grammar.tokens.len() + 1);

  for (i, item) in items.iter().enumerate() {
    if let (prod, 0) = decode_item(max_nsym_p1, item.key) {
      let nt = grammar.prods[prod].nt;
      if !nt_starts.contains_key(&nt) {
        nt_starts.insert(nt, i);
      }
    }
  }

  let mut i = 0;
  while i < items.len() {
    let (prod, dot) = decode_item(max_nsym_p1, items[i].key);
    let prod = &grammar.prods[prod];
    if dot == prod.symbols.len() {
      i += 1;
      continue;
    }

    if let Symbol::Nonterminal(nt) = &prod.symbols[dot] {
      first.clear();
      crate::first::compute_symbols_first(
        &mut first,
        nt_firsts,
        &prod.symbols[dot + 1..],
        Some(&items[i].lookaheads));

      if let Some(&nt_start) = nt_starts.get(nt) {
        let mut changed = false;
        for j in nt_start .. nt_start + grammar.nts[nt].range.len() {
          changed |= items[j].lookaheads.union_with(&first);
        }

        if changed {
          if i > nt_start {
            i = nt_start;
          }
        } else {
          i += 1;
        }
      } else {
        nt_starts.insert(*nt, items.len());

        for prod_ix in grammar.nts[nt].range.clone() {
          items.push(Item {
            key: encode_item(max_nsym_p1, prod_ix, 0),
            lookaheads: first.clone(),
          });
        }

        i += 1;
      }
    } else {
      i += 1;
    }
  }
}

fn compute_transitions(
  grammar: &LoweredGrammar,
  max_nsym_p1: usize,
  state: &State,
) -> Map<Symbol, Vec<Item>> {
  let mut transitions = Map::<_, Vec<Item>>::default();

  for item in &state.items {
    let (prod_ix, dot) = decode_item(max_nsym_p1, item.key);
    let prod = &grammar.prods[prod_ix];
    if dot == prod.symbols.len() {
      continue;
    }

    let next_item = encode_item(max_nsym_p1, prod_ix, dot + 1);
    transitions.entry(prod.symbols[dot].clone())
      .or_default()
      .push(Item {
        key: next_item,
        lookaheads: item.lookaheads.clone(),
      });
  }

  transitions
}

fn encode_item(
  max_nsym_p1: usize,
  prod_ix: usize,
  dot: usize,
) -> u32 {
  (prod_ix * max_nsym_p1 + dot) as u32
}

fn decode_item(
  max_nsym_p1: usize,
  key: u32,
) -> (usize, usize) {
  let prod = key as usize / max_nsym_p1;
  let dot = key as usize % max_nsym_p1;
  (prod, dot)
}

/// Generates ACTION table and GOTO table.
/// 
/// entry in ACTION table:
/// - positive: shift
/// - negative: reduce
/// - zero: error
///
/// entry in GOTO table:
/// - positive: goto
/// - zero: error
pub fn gen_tables<T: LrComputation>(
  builder: &Builder<T>,
) -> Result<(Vec<Vec<i32>>, Vec<Vec<u32>>), Vec<Error>> {
  let num_states = builder.states.len();
  let mut action = vec![vec![0i32; builder.eof as usize + 1]; num_states];
  let mut goto = vec![vec![0u32; builder.grammar.nts.len()]; num_states];
  let mut errors = vec![];

  for (from_state, (_, state)) in builder.states.iter().enumerate() {
    for item in &state.items {
      //let item = builder.item_store.items[item_ix];
      let (prod, dot) = decode_item(builder.max_nsym_p1, item.key);
      let symbols = &builder.grammar.prods[prod].symbols;
      // shift
      if dot < symbols.len() {
        let sym = &symbols[dot];
        let to_state = state.transitions[sym];
        match sym {
          Symbol::Token(tok) => {
            let old = &mut action[from_state][tok.index()];
            if *old < 0 {
              match resolve_sr_conflict(&builder.grammar, !*old as u32, tok.id()) {
                SrConflictResolution::Shift => *old = to_state as i32 + 1,
                SrConflictResolution::Reduce => {
                  // do nothing
                }
                SrConflictResolution::Error => *old = 0,
                SrConflictResolution::Conflict => {
                  errors.push(make_sr_conflict_error(
                    builder,
                    &state.items,
                    tok.id(),
                    !*old as u32));
                }
              }
            } else {
              assert!(*old == 0 || *old == to_state as i32 + 1);
              *old = to_state as i32 + 1;
            }
          }
          Symbol::Nonterminal(nt) => {
            goto[from_state][nt.index()] = to_state + 1;
          }
        }
      } else {
        // reduce
        for lookahead in item.lookaheads.iter() {
          let old = &mut action[from_state][lookahead as usize];
          if *old > 0 {
            match resolve_sr_conflict(&builder.grammar, prod as u32, lookahead) {
              SrConflictResolution::Shift => {
                // do nothing
              }
              SrConflictResolution::Reduce => {
                *old = !(prod as i32);
              }
              SrConflictResolution::Error => {
                *old = 0;
              }
              SrConflictResolution::Conflict => {
                errors.push(make_sr_conflict_error(
                  builder,
                  &state.items,
                  lookahead as u32,
                  prod as u32));
              }
            }
          } else if *old < 0 {
            errors.push(make_rr_conflict_error(
              builder,
              &state.items,
              lookahead as u32,
              !*old as u32,
              prod as u32));
          } else {
            *old = !(prod as i32);
          }
        }
      }
    }
  }

  if errors.is_empty() {
    Ok((action, goto))
  } else {
    Err(errors)
  }
}

enum SrConflictResolution {
  Shift,
  Reduce,
  Error,
  Conflict,
}

fn resolve_sr_conflict(
  grammar: &LoweredGrammar,
  prod_ix: u32,
  tok: u32,
) -> SrConflictResolution {
  match (&grammar.prods[prod_ix as usize].prec, grammar.token_precs.get(&tok)) {
    (Some(prec1), Some((assoc, prec2))) => {
      if prec1 == prec2 {
        match assoc {
          Assoc::LeftAssoc => SrConflictResolution::Reduce,
          Assoc::RightAssoc => SrConflictResolution::Shift,
          Assoc::NonAssoc => SrConflictResolution::Error,
        }
      } else if prec1 < prec2 {
        SrConflictResolution::Shift
      } else {
        SrConflictResolution::Reduce
      }
    }
    _ => SrConflictResolution::Conflict,
  }
}

fn make_rr_conflict_error<T: LrComputation>(
  builder: &Builder<T>,
  item_set: &[Item],
  lookahead: u32,
  reduce1: u32,
  reduce2: u32,
) -> Error {
  let lookahead = builder.grammar.tokens[&lookahead].clone();
  let reduce1 = builder.grammar.prods[reduce1 as usize].to_string(
    &builder.grammar);
  let reduce2 = builder.grammar.prods[reduce2 as usize].to_string(
    &builder.grammar);

  let state_items = item_set.iter()
    .map(|item| {
      let mut buf = String::new();
      builder.fmt_item(item, &mut buf).unwrap();
      buf
    })
    .collect();

  Error::ReduceReduceConflict(ReduceReduceConflictError {
    lookahead,
    state_items,
    reduce1,
    reduce2,
  })
}

fn make_sr_conflict_error<T: LrComputation>(
  builder: &Builder<T>,
  item_set: &[Item],
  token: u32,
  reduce_prod: u32,
) -> Error {
  let shift = builder.grammar.tokens[&token].clone();

  let state_items = item_set.iter()
    .map(|item| {
      let mut buf = String::new();
      builder.fmt_item(item, &mut buf).unwrap();
      buf
    })
    .collect();

  let reduce = builder.grammar.prods[reduce_prod as usize].to_string(&builder.grammar);

  Error::ShiftReduceConflict(ShiftReduceConflictError {
    state_items,
    shift,
    reduce,
  })
}

impl State {
  pub fn new(items: KernelItemSet) -> Self {
    State {
      kernel_len: items.len(),
      items,
      transitions: Map::default(),
    }
  }
}

impl<'a, T: LrComputation> Builder<'a, T> {
  pub fn new(
    grammar: &'a LoweredGrammar
  ) -> Self {
    Self {
      grammar,
      states: StateStore::default(),
      eof: grammar.tokens.len(),
      max_nsym_p1: grammar.prods.iter()
        .map(|prod| prod.symbols.len())
        .max()
        .unwrap()
        + 1,
    }
  }
}

#[cfg(test)]
impl<'a, T: LrComputation> Builder<'a, T> {
  pub fn states(
    &self,
    entry_points: &HashMap<String, EntryPoint>
  ) -> String {
    let mut output = String::new();
    self.fmt_states(entry_points, &mut output).unwrap();
    output
  }

  fn fmt_states(
    &self,
    entry_points: &HashMap<String, EntryPoint>,
    fmt: &mut impl Write,
  ) -> fmt::Result {
    for i in 0..self.states.len() {
      self.fmt_state(entry_points, i, fmt)?;
    }

    Ok(())
  }

  fn fmt_state(
    &self,
    entry_points: &HashMap<String, EntryPoint>,
    state_ix: usize,
    fmt: &mut impl Write,
  ) -> fmt::Result {
    let state = &self.states[state_ix];
    write!(fmt, "State {}", state_ix)?;
    if entry_points.values().find(|x| x.start_state == state_ix as u32).is_some() {
      write!(fmt, " (start)")?;
    }
    writeln!(fmt)?;

    for item in &state.items {
      self.fmt_item(item, fmt)?;

      writeln!(fmt)?;
    }

    writeln!(fmt)
  }
}

impl<'a, T: LrComputation> Builder<'a, T> {
  fn fmt_item(
    &self,
    item: &Item,
    fmt: &mut impl Write,
  ) -> fmt::Result {
    //let item = self.item_store.items[item_ix];
    let (prod, dot) = decode_item(self.max_nsym_p1, item.key);
    let nt = self.grammar.prods[prod].nt;
    let symbols = &self.grammar.prods[prod].symbols;

    write!(fmt, "{} ->", self.grammar.nts[&nt].name)?;

    for (i, sym) in symbols.iter().enumerate() {
      if i == dot {
        write!(fmt, " .")?;
      }

      match sym {
        Symbol::Token(token) => {
          let name = self.grammar.tokens.get(token).map(|s|s.as_str()).unwrap_or("$");
          write!(fmt, " {}", name)?;
        }
        Symbol::Nonterminal(nt) => {
          let name = &self.grammar.nts[nt].name;
          write!(fmt, " {}", name)?;
        }
      }
    }

    if dot == symbols.len() {
      write!(fmt, " .")?;
    }

    write!(fmt, "      ")?;

    let mut slash = false;
    for lookahead in item.lookaheads.iter() {
      if slash {
        write!(fmt, " / ")?;
      }
      slash = true;

      write!(fmt, "{}",
        self.grammar.tokens.get(&lookahead).map(|s|s.as_str()).unwrap_or("$"))?;
    }

    Ok(())
  }
}