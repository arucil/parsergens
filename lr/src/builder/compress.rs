use grammar::{LoweredGrammar, HashMap};
use std::hash::Hash;
use num_traits::int::PrimInt;

#[derive(Debug, Default)]
pub struct CompressedTables {
  /// ACTION table + GOTO table
  pub parse_table: Vec<i32>,

  /// displacement values for rows of ACTION table.
  pub action_disp: Vec<isize>,

  /// displacement values for rows of GOTO table.
  pub goto_disp: Vec<isize>,

  /// default values for rows of ACTION table.
  pub action_default: Vec<i32>,

  /// default values for rows of GOTO table.
  pub goto_default: Vec<i32>,

  /// -1 in `check_table` represents an empty entry.
  pub check_table: Vec<i32>,
}

#[derive(Debug, Default)]
struct CompressState {
  parse_table: Vec<i32>,
  check_table: Vec<i32>,
  first_empty: usize,
}

pub fn compress_tables(
  grammar: &LoweredGrammar,
  action: Vec<Vec<i32>>,
  goto: Vec<Vec<u32>>,
) -> CompressedTables {
  let mut state = CompressState::default();
  let num_tokens = grammar.tokens.len() + 1; // includes EOF

  let (action_disp, action_default) = compress_action_table(
    &mut state,
    action,
  );

  let (goto_disp, goto_default) = compress_goto_table(
    &mut state,
    goto,
    num_tokens,
  );

  while state.check_table.last() == Some(&-1) {
    state.check_table.pop();
    state.parse_table.pop();
  }

  for x in &mut state.parse_table {
    if *x == super::tables::ERROR_ACTION {
      *x = 0;
    }
  }

  CompressedTables {
    parse_table: state.parse_table,
    check_table: state.check_table,
    action_disp,
    action_default,
    goto_disp,
    goto_default,
  }
}

/// Returns displacement table and default value table.
fn compress_action_table(
  state: &mut CompressState,
  action: Vec<Vec<i32>>,
) -> (Vec<isize>, Vec<i32>) {
  let mut action = action.into_iter()
    .enumerate()
    .map(|(i, row)| make_row(i, row, |x| x > 0 || x == super::tables::ERROR_ACTION))
    .collect::<Vec<_>>();

  let default_table = action.iter().map(|row| row.default).collect::<Vec<_>>();

  action.sort_by_key(|row| row.num_default);
  let action = action;

  let mut disp_map = HashMap::<&[i32], isize>::default();
  let mut disp_table = vec![0; action.len()];

  for row in &action {
    if let Some(&disp) = disp_map.get(&row.values[..]) {
      // merge rows with exactly the same actions.
      disp_table[row.index] = disp;
    } else {
      // check_base is token id
      let disp = write_row(state, row, 0, false, true);
      disp_map.insert(&row.values, disp);
      disp_table[row.index] = disp;
    }
  }

  (disp_table, default_table)
}

/// Returns displacement table and default value table.
fn compress_goto_table(
  state: &mut CompressState,
  goto: Vec<Vec<u32>>,
  num_tokens: usize,
) -> (Vec<isize>, Vec<i32>) {
  let mut goto = goto.into_iter()
    .enumerate()
    .map(|(i, row)| make_row(i, row, |_| false))
    .collect::<Vec<_>>();

  let default_table = goto.iter().map(|row| row.default as i32).collect::<Vec<_>>();

  goto.sort_by_key(|row| row.num_default);
  let goto = goto;

  let mut disp_table = vec![0; goto.len()];

  for row in &goto {
    // check_base is token NT + num_tokens
    let disp = write_row(state, row, num_tokens, true, false);
    disp_table[row.index] = disp;
  }

  (disp_table, default_table)
}

/// Returns the displacement.
fn write_row<T: PrimInt>(
  state: &mut CompressState,
  row: &Row<T>,
  check_base: usize,
  check_row: bool,
  merge_rows: bool,
) -> isize {
  let mut disp = state.first_empty as isize;

  for &v in &row.values {
    if v == T::zero() {
      disp -= 1;
    } else {
      break;
    }
  }

  if state.first_empty as isize - disp == row.values.len() as isize {
    return -(row.values.len() as isize);
  }

  if merge_rows {
    'find_disp: loop {
      let mut i = disp;
      for (j, &v) in row.values.iter().enumerate() {
        if i >= 0 && (i as usize) < state.check_table.len() &&
          state.check_table[i as usize] >= 0
        {
          let i = i as usize;
          let next = if v == T::zero() {
            // two rows that have unoverlapped non-zero values are not allowed
            // to be merged.
            state.check_table[i] == j as i32
          } else {
            state.check_table[i] != j as i32 || state.parse_table[i] != v.to_i32().unwrap()
          };
          if next {
            disp += 1;
            continue 'find_disp;
          }
        }
        i += 1;
      }
      break;
    }
  } else {
    'find_disp2: loop {
      let mut i = disp;
      for &v in &row.values {
        if !(v == T::zero() || i < 0 ||
          (i as usize) >= state.check_table.len() ||
          state.check_table[i as usize] < 0)
        {
          disp += 1;
          continue 'find_disp2;
        }
        i += 1;
      }
      break;
    }
  }

  let row_end = (disp + row.values.len() as isize) as usize;
  if row_end > state.check_table.len() {
    state.check_table.resize(row_end, -1);
    state.parse_table.resize(row_end, 0);
  }

  for (i, &v) in row.values.iter().enumerate() {
    if v != T::zero() {
      let j = (disp + i as isize) as usize;
      assert!(state.check_table[j] == -1);
      state.parse_table[j] = v.to_i32().unwrap();
      state.check_table[j] = if check_row {
        (check_base + row.index) as i32
      } else {
        (check_base + i) as i32
      };
    }
  }

  while state.first_empty < state.check_table.len() &&
    state.check_table[state.first_empty] == -1
  {
    state.first_empty += 1;
  }

  disp
}

fn make_row<T, F>(
  index: usize,
  mut values: Vec<T>,
  ignore_value: F,
) -> Row<T>
where
  T: Hash + PrimInt,
  F: Fn(T) -> bool,
{
  let mut count = HashMap::<T, usize>::default();
  let mut num_zeros = 0;

  for &v in &values {
    if v == T::zero() {
      num_zeros += 1;
    } else if !ignore_value(v) {
      *count.entry(v).or_default() += 1;
    }
  }

  if let Some((default, num_default)) = count.into_iter().max_by_key(|x| x.1) {
    // clear default values
    for v in &mut values {
      if *v == default {
        *v = T::zero();
      }
    }

    Row {
      index,
      default,
      num_default: num_default + num_zeros,
      values,
    }
  } else {
    Row {
      index,
      default: T::zero(),
      num_default: num_zeros,
      values,
    }
  }
}

#[derive(Debug, PartialEq, Eq)]
struct Row<T> {
  index: usize,
  default: T,
  num_default: usize,
  values: Vec<T>,
}

#[cfg(test)]
mod tests {
  use pretty_assertions::assert_eq;
  use super::*;

  #[test]
  fn make_row_all_zeros() {
    let vec = vec![0; 13];
    let row = make_row(1, vec.clone(), |_| false);
    assert_eq!(row, Row {
      index: 1,
      default: 0,
      num_default: 13,
      values: vec,
    });
  }

  #[test]
  fn make_row_most_frequent() {
    let vec1 = vec![0, 0, 7, 2, 0, 13, 7, 2, 0, 7, 11];
    let vec2 = vec![0, 0, 0, 2, 0, 13, 0, 2, 0, 0, 11];
    let row = make_row(1, vec1, |_| false);
    assert_eq!(row, Row {
      index: 1,
      default: 7,
      num_default: 7,
      values: vec2,
    });
  }

  #[test]
  fn write_rows() {
    let mut state = CompressState::default();
    let row1 = make_row(1, vec![0, 0, 0, 2, 3, 2, 0], |_| false);

    assert_eq!(
      -4,
      write_row(&mut state, &row1, 10, false, true)
    );

    assert_eq!(
      vec![3, 0, 0],
      state.parse_table,
    );

    assert_eq!(
      vec![14, -1, -1],
      state.check_table,
    );

    let row2 = make_row(1, vec![5, 8, 5, 0, -3, 0, 2, -1], |_| false);

    assert_eq!(
      0,
      write_row(&mut state, &row2, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 0, 0, -3, 0, 2, -1],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, -1, -1, 14, -1, 16, 17],
      state.check_table,
    );

    let row3 = make_row(1, vec![0; 8], |_| false);

    assert_eq!(
      -8,
      write_row(&mut state, &row3, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 0, 0, -3, 0, 2, -1],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, -1, -1, 14, -1, 16, 17],
      state.check_table,
    );

    let row4 = make_row(1, vec![3, 1, 1, 6, 7, 1, 1, 55], |_| false);

    assert_eq!(
      5,
      write_row(&mut state, &row4, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 0, 0, -3, 3, 2, -1, 6, 7, 0, 0, 55],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, -1, -1, 14, 10, 16, 17, 13, 14, -1, -1, 17],
      state.check_table,
    );

    let row5 = make_row(1, vec![0, 0, 12, 6, 5, 5, 0, 5], |_| false);

    assert_eq!(
      0,
      write_row(&mut state, &row5, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 12, 6, -3, 3, 2, -1, 6, 7, 0, 0, 55],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, 12, 13, 14, 10, 16, 17, 13, 14, -1, -1, 17],
      state.check_table,
    );

    let row6 = make_row(1, vec![1, 2, 3, 0, 4, 4, 0, 0], |_| false);

    assert_eq!(
      13,
      write_row(&mut state, &row6, 10, false, true)
    );

    assert_eq!(
      vec![3, 8, 12, 6, -3, 3, 2, -1, 6, 7, 0, 0, 55, 1, 2, 3, 0, 0, 0, 0, 0],
      state.parse_table,
    );

    assert_eq!(
      vec![14, 11, 12, 13, 14, 10, 16, 17, 13, 14, -1, -1, 17, 10, 11, 12, -1, -1, -1, -1, -1],
      state.check_table,
    );
  }

  #[test]
  fn action_table() {
    let mut state = CompressState::default();

    let (disp, default) = compress_action_table(
      &mut state,
      vec![
        vec![0, 1, -4, -4, 2],
        vec![3, 4, 0, 5, 0],
        vec![6, 0, 7, 0, -12],
        vec![0, 0, 0, 8, 0],
        vec![9, -3, 0, 0, 10],
      ],
    );

    assert_eq!(
      state.parse_table,
      vec![3, 4, 1, 5, 6, 2, 7, 9, 8, 0, 0, 10],
    );

    assert_eq!(
      state.check_table,
      vec![0, 1, 1, 3, 0, 4, 2, 0, 3, -1, -1, 4],
    );

    assert_eq!(
      disp,
      vec![1, 0, 4, 5, 7],
    );

    assert_eq!(
      default,
      vec![-4, 0, -12, 0, -3],
    );
  }

  #[test]
  fn goto_table() {
    let mut state = CompressState::default();

    let (disp, default) = compress_goto_table(
      &mut state,
      vec![
        vec![0, 1, 4, 4, 2],
        vec![3, 4, 9, 5, 9],
        vec![6, 12, 7, 0, 12],
        vec![0, 1, 1, 8, 0],
        vec![9, 3, 0, 3, 10],
      ],
      100
    );

    assert_eq!(
      state.parse_table,
      vec![3, 4, 1, 5, 6, 2, 7, 9, 8, 0, 0, 10],
    );

    assert_eq!(
      state.check_table,
      vec![101, 101, 100, 101, 102, 100, 102, 104, 103, -1, -1, 104],
    );

    assert_eq!(
      disp,
      vec![1, 0, 4, 5, 7],
    );

    assert_eq!(
      default,
      vec![4, 9, 12, 1, 3],
    );
  }

  #[test]
  fn action_table_duplicate_rows() {
    let mut state = CompressState::default();

    let (disp, default) = compress_action_table(
      &mut state,
      vec![
        vec![0, 1, -4, -4, 2],
        vec![3, 4, 0, 5, 0],
        vec![6, 0, 7, 0, -12],
        vec![0, 0, 0, 8, 0],
        vec![9, -3, 0, 0, 10],
        vec![9, 0, 0, -1, 10],
        vec![9, 0, 0, 0, 0],
      ],
    );

    assert_eq!(
      state.parse_table,
      vec![3, 4, 1, 5, 6, 2, 7, 9, 8, 9, 0, 10, 0, 0],
    );

    assert_eq!(
      state.check_table,
      vec![0, 1, 1, 3, 0, 4, 2, 0, 3, 0, -1, 4, -1, -1],
    );

    assert_eq!(
      disp,
      vec![1, 0, 4, 5, 7, 7, 9],
    );

    assert_eq!(
      default,
      vec![-4, 0, -12, 0, -3, -1, 0],
    );
  }
}