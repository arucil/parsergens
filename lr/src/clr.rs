use crate::builder::{LrComputation, StateStore, KernelItemSet, State, Item};

pub struct ClrComputation;

impl LrComputation for ClrComputation {
  type StateKey = Vec<Item>;

  /// `kernel_item_set` is sorted by the field `key`.
  ///
  /// Returns whether the state has changed.
  fn store_state(
    states: &mut StateStore<Self::StateKey>,
    kernel_item_set: KernelItemSet,
  ) -> (u32, bool) {
    if let Some(i) = states.get_index_of(&kernel_item_set) {
      (i as u32, false)
    } else {
      let state_ix = states.insert_full(
        kernel_item_set.clone(),
        State::new(kernel_item_set),
      ).0 as u32;

      (state_ix, true)
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::builder::*;
  use insta::{assert_snapshot, assert_debug_snapshot};
  use grammar::LoweredGrammar;
  use crate::augment;

  fn prepare(input: &str) -> LoweredGrammar {
    let grammar = grammar::build(input).unwrap();
    let grammar = grammar.lower();
    let grammar = augment::augment(grammar);

    grammar
  }

  fn merge_action_goto((action, goto): (Vec<Vec<i32>>, Vec<Vec<u32>>)) -> Vec<Vec<i32>> {
    action.into_iter().enumerate().map(|(state, mut row1)| {
      row1.extend((0..goto.len()).map(|nt| goto[nt][state] as i32));
      row1
    })
    .collect()
  }

  static SIMPLE: &str = r#"
%token c "c"
%token d "d"
%start S
S = C C
C = c C
  | d
  "#;

  #[test]
  fn simple_states() {
    let grammar = prepare(SIMPLE);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    let start_nts = gen_states(&mut builder);

    assert_snapshot!(builder.states(&start_nts));
  }

  #[test]
  fn simple_action_goto() {
    let grammar = prepare(SIMPLE);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    gen_states(&mut builder);
    let tables = merge_action_goto(gen_tables(&builder).unwrap());

    assert_debug_snapshot!(tables);
  }

  static EPSILON: &str = r#"
%token plus "+"
%token mult "*"
%token num "1"
%token lparen "("
%token rparen ")"

%start E

E = T E'
E' = plus T E'
   | ()
T = F T'
T' = mult F T'
   | ()
F = num
  | lparen E rparen
  "#;

  #[test]
  fn epsilon_states() {
    let grammar = prepare(EPSILON);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    let start_nts = gen_states(&mut builder);

    assert_snapshot!(builder.states(&start_nts));
  }

  #[test]
  fn epsilon_action_goto() {
    let grammar = prepare(EPSILON);
    let mut builder = Builder::<ClrComputation>::new(&grammar);
    let _start_nts = gen_states(&mut builder);
    let tables = gen_tables(&builder).unwrap();
    let tables = merge_action_goto(tables);

    assert_debug_snapshot!(tables);
  }
}