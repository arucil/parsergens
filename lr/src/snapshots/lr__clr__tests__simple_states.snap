---
source: lr/src/clr.rs
expression: builder.states()
---
State 0 (start)
_S -> . S      $
S -> . A a      $
S -> . b A c      $
S -> . d c      $
S -> . b d a      $
A -> . d      a

State 1
_S -> S .      $

State 2
S -> A . a      $

State 3
S -> b . A c      $
S -> b . d a      $
A -> . d      c

State 4
S -> d . c      $
A -> d .      a

State 5
S -> A a .      $

State 6
S -> b A . c      $

State 7
S -> b d . a      $
A -> d .      c

State 8
S -> d c .      $

State 9
S -> b A c .      $

State 10
S -> b d a .      $


